rosserial_server/include/rosserial_server/topic_handlers.h

```diff
@@ -184,17 +184,117 @@ class ServiceClient {
   boost::function<void(std::vector<uint8_t>& buffer, const uint16_t topic_id)> write_fn_;
   std::string service_md5_;
   std::string request_message_md5_;
   std::string response_message_md5_;
   uint16_t topic_id_;
 };
 
-ros::ServiceClient ServiceClient::service_info_service_;
 typedef boost::shared_ptr<ServiceClient> ServiceClientPtr;
 
+
+class ServiceServer {
+public:
+  ServiceServer(ros::NodeHandle& nh, rosserial_msgs::TopicInfo& topic_info,
+      boost::function<void(std::vector<uint8_t>& buffer, const uint16_t topic_id)> write_fn)
+    : write_fn_(write_fn) {
+    topic_id_ = -1;
+    get_response_ = false;
+
+    rosserial_server::MsgInfo srvinfo;
+    rosserial_server::MsgInfo reqinfo;
+    rosserial_server::MsgInfo respinfo;
+    try
+    {
+      srvinfo = lookupMessage(topic_info.message_type, "srv");
+      reqinfo = lookupMessage(topic_info.message_type + "Request", "srv");
+      respinfo = lookupMessage(topic_info.message_type + "Response", "srv");
+    }
+    catch (const std::exception& e)
+    {
+      ROS_WARN_STREAM("Unable to look up service definition: " << e.what());
+    }
+
+    service_md5_ = srvinfo.md5sum;
+    request_message_md5_ = reqinfo.md5sum;
+    response_message_md5_ = respinfo.md5sum;
+
+    ros::AdvertiseServiceOptions opts;
+    opts.service = topic_info.topic_name;
+    opts.md5sum = service_md5_;
+    opts.datatype = topic_info.message_type;
+    opts.req_datatype = topic_info.message_type + "Request";
+    opts.res_datatype = topic_info.message_type + "Response";
+    opts.helper = boost::make_shared<ros::ServiceCallbackHelperT<ros::ServiceSpec<topic_tools::ShapeShifter,topic_tools::ShapeShifter> > >(boost::bind(&ServiceServer::request_handle, this, _1, _2));
+
+    service_server_ = nh.advertiseService(opts);
+  }
+  void setTopicId(uint16_t topic_id) {
+    topic_id_ = topic_id;
+  }
+  std::string getServiceMD5() {
+    return service_md5_;
+  }
+  std::string getRequestMessageMD5() {
+    return request_message_md5_;
+  }
+  std::string getResponseMessageMD5() {
+    return response_message_md5_;
+  }
+
+  bool request_handle(topic_tools::ShapeShifter& request_message, topic_tools::ShapeShifter& response_message) {
+
+    get_response_ = false;
+    ROS_ERROR_STREAM("service receive " << service_server_.getService() << " request");
+
+    size_t length = ros::serialization::serializationLength(request_message);
+    std::vector<uint8_t> buffer(length);
+
+    ros::serialization::OStream ostream(&buffer[0], length);
+    ros::serialization::Serializer<topic_tools::ShapeShifter>::write(ostream, request_message);
+    write_fn_(buffer,topic_id_);
+
+    //wait for the response
+    int cnt = 0;
+    while(!get_response_)
+      {
+        ros::Duration(0.01).sleep();
+
+        if (cnt++ > 500) // hard-coding
+          {
+            ROS_WARN_STREAM(service_server_.getService() << ": no response!!");
+            return false;
+          }
+      }
+
+    response_message = response_message_;
```

Comment:
Hi, I am getting segfault at this line after I call a custom service at the third time. Any tip how to fix it? I comment this code line and the segfault disappears, but consequently, I do not receive the response message.

Reply:
@rezeck

Sorry for the very late response. 
I have tested  with https://github.com/ros-drivers/rosserial/blob/noetic-devel/rosserial_arduino/src/ros_lib/examples/ServiceServer/ServiceServer.pde, however I can not encounter the problem you pointed out. Could you paste your code to let me test.

By the way, as @romainreignier pointed out,  there is no  synchronization primitives to protect the resources in this commit, which might be the reason of  your problem. I will add mutex in multiple threads.

Reply:
I have carefully studied this problem again, found there is a fatal bug due to `response_message = response_message_;`
`response_message` which the type is [topic_tools::ShapeShifter](http://docs.ros.org/kinetic/api/topic_tools/html/classtopic__tools_1_1ShapeShifter.html), has a pointer type member `uint8_t * msgBuf`. 

 It seems like the `response_message` will be deleted after this rosservice server callback function, which leads to a invalid access to `msgBuf` in   `response_message_`, in next rosservice response.

I have fixed this problem by commit  3ffc8e4 . Could you please help me to check the latest version of the branch?